### PR TITLE
Test Notebooks on CI

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?
Adopts JAX's `myst_nb` + `jupytext` workflow.
### Tasks
- [x] Uses `myst_nb` to render notebooks instead of `nbsphinx`. `myst_nb` executes the notebooks before rendering therefore it also serves as a testing mechanism.
- [x] Adds the `jupytext` pre-commit hook and pre-commit CI job to keep companion `.md` files for each notebook, these mardown files are kept in sync and can be easily review on PRs.
- [x] Revive "Open In Colab" buttons

### CODE REVIEW

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Really nice and useful change @cgarciae! Few questions:
- Did you measure the impact on our CI runtime? I am curious how much longer it takes.
- IIUC then the generated .md files are not used for anything else except for simplifying modification of the notebooks, right? I wonder if at some point we can instead generate .rst files from them, then we could have our ideal situation of only having one source of truth for our HOWTOs.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -13,6 +13,22 @@ on:
...
+      - name: Cancel previous
+        uses: styfle/cancel-workflow-action@0.9.1
+        with:
+          access_token: ${{ github.token }}
+        if: ${{github.ref != 'refs/head/main'}}
+      - uses: actions/checkout@v3
+      - name: Set up Python 3.8
+        uses: actions/setup-python@v3
+        with:
+          python-version: 3.8
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply .github/workflows/build.yml:30]: Yes, good catch!

File: docs/contributing.md [hunk 2]
```diff
@@ -109,6 +109,73 @@ Follow these steps to contribute code:
    Consult [GitHub Help](https://help.github.com/articles/about-pull-requests/)
    for more information on using pull requests.
 
+### Update notebooks
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply docs/contributing.md:112]: This is ported from JAX docs, creddit goes to the JAX team :slightly_smiling_face:

File: docs/contributing.md [hunk 3]
```diff
@@ -109,6 +109,73 @@ Follow these steps to contribute code:
...
+
+The jupytext version should match that specified in
+[.pre-commit-config.yaml](https://github.com/google/flax/blob/main/.pre-commit-config.yaml).
+
+To check that the markdown and ipynb files are properly synced, you may use the
+[pre-commit](https://pre-commit.com/) framework to perform the same check used
+by the github CI:
+
+```
+git add docs -u  # pre-commit runs on files in git staging.
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply docs/contributing.md:151]: Added a section above with instructions on how to setup `pre-commit` as this will now be part of the main workflow.

**Comment by cgarciae** [comment]
@marcvanzee 

> Did you measure the impact on our CI runtime? I am curious how much longer it takes.

Will get the numbers but for sure this will be slower :sweat_smile:, each cell is capped to a 100 seconds runtime though.

> IIUC then the generated .md files are not used for anything else except for simplifying modification of the notebooks, right? I wonder if at some point we can instead generate .rst files from them, then we could have our ideal situation of only having one source of truth for our HOWTOs.

HOWTOs that don't use any fancy sphinx feature can be written in notebook/markdown, no need to generate an `.rst` file since `myst_nb` will render it in html for us. I believe the motivation for the current setup is the `.. testsetup::` directive that hides some code while keeping the HOWTO executable right? `myst_nb`'s `hide-cell` and `remove-cell` tags can accommodate for this. I'll create an issue for this.

**Comment by cgarciae** [comment]
@marcvanzee Here are the numbers, it add ~125 seconds. Note that `jax_for_the_impatient` is excluded for now. I think #2184 could help in this regard.

| | nbsphinx | myst_nb
|--|--|--|
|  time (secs) | 44.70 | 161.13 |

**Comment by cgarciae** [comment]
> We also have the side-by-side diff view in our HOWTOs, I suppose this can't be used anymore then?

Yeah, I imagine it will not be easy to port those, there are a couple like the `model_surgery` HOWTO that don't use that mechanic which we could port right now.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+# 'pre-commit autoupdate'
+
+# Run the hooks on all files with
+# 'pre-commit run --all'
+
+- repo: https://github.com/mwouts/jupytext
+  rev: v1.13.8
+  hooks:
+  - id: jupytext
+    args: [--sync]
```
Comment (marcvanzee) [inline .pre-commit-config.yaml:14]:
```suggestion
    args: [--sync]

```

File: docs/contributing.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply docs/contributing.md:159]: Thanks!

**Comment by marcvanzee** [comment]
Looks good! The notebooks show a huge amount of diffs, but you only added the "Show in Colab" button right? It is so hard to review notebook changes in `.ipynb` format, which is another good argument for this change! 😄

**Review by marcvanzee — APPROVED** [review]

**Comment by cgarciae** [comment]
@marcvanzee yeah git has a hard time with these large json files, hopefully we start relying on the markdown files from now on 😃

**Pull request merged**